packages/graphiql-build-fetcher/package.json

```diff
@@ -0,0 +1,33 @@
+{
+  "name": "@graphiql/build-fetcher",
+  "version": "0.0.1",
+  "description": "Utility to build a fetcher for GraphiQL",
+  "contributors": [
+    "Rikki Schulte <rikki.schulte@gmail.com> (http://rikki.dev/"
+  ],
+  "repository": "http://github.com/graphql/graphiql",
+  "homepage": "http://github.com/graphql/graphiql/tree/master/packages/graphiql#readme",
+  "bugs": {
+    "url": "https://github.com/graphql/graphiql/issues?q=issue+label:graphiql"
+  },
+  "license": "MIT",
+  "main": "dist/index.js",
+  "module": "esm/index.js",
+  "scripts": {
+    "server": "ts-node test/server.ts"
+  },
+  "dependencies": {
+    "graphql-ws": "^4.1.0",
+    "graphql-transport-ws": "^1.9.0",
```

Comment:
Why are you using the deprecated [`graphql-transport-ws`](https://www.npmjs.com/package/graphql-transport-ws) lib? Its just an old name for `graphql-ws`, the two are fundamentally the same. [`graphql-ws@1.9.1`](https://github.com/enisdenjo/graphql-ws/releases/tag/v1.9.1) = [`graphql-transport-ws@1.9.0`](https://github.com/enisdenjo/graphql-ws/releases/tag/v1.9.0)

I wouldnt advocate the usage of the deprecated library at all. The full focus should be on `graphql-ws`.

```suggestion
```

Reply:
we want to have backwards compatibility with the old transport spec since it's so widely used?

Reply:
Thats [`subscriptions-transport-ws`](https://github.com/apollostack/subscriptions-transport-ws).

Reply:
is `graphql-ws` backwards compatible with older client implementations?

Reply:
[`graphql-ws` implements the new Protocol exclusively.](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md) It is not interoperable with [Apollo's Protocol](https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md).

A WebSocket can use **just one subprotocol**; therefore, built-in interoperability is simply not feasible. Read more about this topic [here](https://github.com/enisdenjo/graphql-ws/issues/83) and [here](https://github.com/graphql/graphql-over-http/pull/140#issuecomment-739823000).

If you want to support both the deprecated Apollo's lib and the new one, [you have to configure a connection delegator on the server side](https://github.com/enisdenjo/graphql-ws#ws-backwards-compat).

Reply:
@benjie what is your take on this? from what I understand, we can just attempt a connection with `graphql-ws` protocol and if it fails, fail over to apollo's protocol. that's what we are doing here in this PR. @enisdenjo the reality here is that most of our users using websockets will probably still be using the old protocol for now and not want to change just for graphiql, even though it's being deprecated. we could release with only `graphql-ws` support, but it will probably lead to a flurry of bug reports and confusion even if we document this. 

that said, pointing out that the apollo protocol is deprecated might be enough, but we have users using the old apollo protocol in many other language implementations still.

ad-hoc standards are so confusing, good thing you're finalizing all of this into an adoptable standard. maybe the hard luck approach is best, and if they need backwards compatibility

Reply:
Yes, thats correct. We try using `graphql-ws` and if the socket closes with the close event: `1002: Protocol Error` its very likely that the subprotocol is denied ([thats how Apollo's WS server denies invalid subprotocols](https://github.com/apollographql/subscriptions-transport-ws/blob/3590e076f549744af717dd3257fce78a659b7176/src/server.ts#L148)). In this case, fall back to `subscriptions-transport-ws`.

Reply:
Specifically this line of code (`"graphql-transport-ws": "^1.9.0"`) should be removed, @enisdenjo authored both and renamed `graphql-transport-ws` to simply `graphql-ws` :+1: 

---

On the broader subject of `subscriptions-transport-ws` vs `graphql-ws` support; I also agree with @enisdenjo - the best approach is to try connecting with `graphql-ws` and if that connection fails then try again using `subscriptions-transport-ws`. Even though probably 95% of GraphQL APIs that support websockets currently don't support `graphql-ws`, in future this should shift since Apollo's protocol is officially unmaintained; we should skate to where the puck is going not where it has been. It's possible for servers to support both protocols (as the next release of PostGraphile does), in that case `graphql-ws` should win.

Reply:
@benjie awesome! thats what we're doing in this PR already, just convincing @endisdenjo that it makese sense

Reply:
I think what I meant to add here is this:

https://www.npmjs.com/package/subscriptions-transport-ws

got `graphql-transport-ws` and `subscriptions-transport-ws` mixed up sorry!

Reply:
@enisdenjo my apologies for taking so long to realize this 😆  I knew something was wrong when I couldn't import the `SubscriptionsClient` class I was used to!

Reply:
Haha, thats ok! Glad you figured it out. 😄